File: civildesk_frontend/lib/models/salary_slip.dart
class SalarySlip {
  final int? id;
  final String employeeId;
  final String employeeName;
  final String? department;
  final String? designation;
  final int year;
  final int month;
  final String periodString;
  
  // Calendar & Working Days
  final int? totalDaysInMonth;
  final int? workingDays;
  final int? weeklyOffs;
  
  // Attendance Data
  final double? totalEffectiveWorkingHours;
  final double? totalOvertimeHours;
  final double? rawPresentDays;
  final int? presentDays;
  final int? absentDays;
  final double? prorationFactor;
  
  // Earnings
  final double? basicPay;
  final double? hraAmount;
  final double? medicalAllowance;
  final double? conveyanceAllowance;
  final double? uniformAndSafetyAllowance;
  final double? bonus;
  final double? foodAllowance;
  final double? specialAllowance;
  final double? overtimePay;
  final double? totalSpecialAllowance;
  final double? otherIncentive;
  final double? epfEmployerEarnings;
  final double? totalEarnings;
  
  // Deductions
  final double? epfEmployeeDeduction;
  final double? epfEmployerDeduction;
  final double? esicDeduction;
  final double? professionalTax;
  final double? tds;
  final double? advanceSalaryRecovery;
  final double? loanRecovery;
  final double? fuelAdvanceRecovery;
  final double? otherDeductions;
  final double? totalStatutoryDeductions;
  final double? totalOtherDeductions;
  final double? totalDeductions;
  
  // Net Salary
  final double? netSalary;
  
  // Rates
  final double? dailyRate;
  final double? hourlyRate;
  final double? overtimeRate;
  
  // Status
  final String status;
  final int? generatedBy;
  final DateTime? generatedAt;
  final String? notes;
  
  // Metadata
  final DateTime? createdAt;
  final DateTime? updatedAt;

  SalarySlip({
    this.id,
    required this.employeeId,
    required this.employeeName,
    this.department,
    this.designation,
    required this.year,
    required this.month,
    required this.periodString,
    this.totalDaysInMonth,
    this.workingDays,
    this.weeklyOffs,
    this.totalEffectiveWorkingHours,
    this.totalOvertimeHours,
    this.rawPresentDays,
    this.presentDays,
    this.absentDays,
    this.prorationFactor,
    this.basicPay,
    this.hraAmount,
    this.medicalAllowance,
    this.conveyanceAllowance,
    this.uniformAndSafetyAllowance,
    this.bonus,
    this.foodAllowance,
    this.specialAllowance,
    this.overtimePay,
    this.totalSpecialAllowance,
    this.otherIncentive,
    this.epfEmployerEarnings,
    this.totalEarnings,
    this.epfEmployeeDeduction,
    this.epfEmployerDeduction,
    this.esicDeduction,
    this.professionalTax,
    this.tds,
    this.advanceSalaryRecovery,
    this.loanRecovery,
    this.fuelAdvanceRecovery,
    this.otherDeductions,
    this.totalStatutoryDeductions,
    this.totalOtherDeductions,
    this.totalDeductions,
    this.netSalary,
    this.dailyRate,
    this.hourlyRate,
    this.overtimeRate,
    required this.status,
    this.generatedBy,
    this.generatedAt,
    this.notes,
    this.createdAt,
    this.updatedAt,
  });

  factory SalarySlip.fromJson(Map<String, dynamic> json) {
    return SalarySlip(
      id: json['id'],
      employeeId: json['employeeId'] ?? '',
      employeeName: json['employeeName'] ?? '',
      department: json['department'],
      designation: json['designation'],
      year: json['year'] ?? 0,
      month: json['month'] ?? 0,
      periodString: json['periodString'] ?? '',
      totalDaysInMonth: json['totalDaysInMonth'],
      workingDays: json['workingDays'],
      weeklyOffs: json['weeklyOffs'],
      totalEffectiveWorkingHours: json['totalEffectiveWorkingHours']?.toDouble(),
      totalOvertimeHours: json['totalOvertimeHours']?.toDouble(),
      rawPresentDays: json['rawPresentDays']?.toDouble(),
      presentDays: json['presentDays'],
      absentDays: json['absentDays'],
      prorationFactor: json['prorationFactor']?.toDouble(),
      basicPay: json['basicPay']?.toDouble(),
      hraAmount: json['hraAmount']?.toDouble(),
      medicalAllowance: json['medicalAllowance']?.toDouble(),
      conveyanceAllowance: json['conveyanceAllowance']?.toDouble(),
      uniformAndSafetyAllowance: json['uniformAndSafetyAllowance']?.toDouble(),
      bonus: json['bonus']?.toDouble(),
      foodAllowance: json['foodAllowance']?.toDouble(),
      specialAllowance: json['specialAllowance']?.toDouble(),
      overtimePay: json['overtimePay']?.toDouble(),
      totalSpecialAllowance: json['totalSpecialAllowance']?.toDouble(),
      otherIncentive: json['otherIncentive']?.toDouble(),
      epfEmployerEarnings: json['epfEmployerEarnings']?.toDouble(),
      totalEarnings: json['totalEarnings']?.toDouble(),
      epfEmployeeDeduction: json['epfEmployeeDeduction']?.toDouble(),
      epfEmployerDeduction: json['epfEmployerDeduction']?.toDouble(),
      esicDeduction: json['esicDeduction']?.toDouble(),
      professionalTax: json['professionalTax']?.toDouble(),
      tds: json['tds']?.toDouble(),
      advanceSalaryRecovery: json['advanceSalaryRecovery']?.toDouble(),
      loanRecovery: json['loanRecovery']?.toDouble(),
      fuelAdvanceRecovery: json['fuelAdvanceRecovery']?.toDouble(),
      otherDeductions: json['otherDeductions']?.toDouble(),
      totalStatutoryDeductions: json['totalStatutoryDeductions']?.toDouble(),
      totalOtherDeductions: json['totalOtherDeductions']?.toDouble(),
      totalDeductions: json['totalDeductions']?.toDouble(),
      netSalary: json['netSalary']?.toDouble(),
      dailyRate: json['dailyRate']?.toDouble(),
      hourlyRate: json['hourlyRate']?.toDouble(),
      overtimeRate: json['overtimeRate']?.toDouble(),
      status: json['status'] ?? 'DRAFT',
      generatedBy: json['generatedBy'],
      generatedAt: json['generatedAt'] != null ? DateTime.parse(json['generatedAt']) : null,
      notes: json['notes'],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'employeeId': employeeId,
      'employeeName': employeeName,
      'department': department,
      'designation': designation,
      'year': year,
      'month': month,
      'periodString': periodString,
      'totalDaysInMonth': totalDaysInMonth,
      'workingDays': workingDays,
      'weeklyOffs': weeklyOffs,
      'totalEffectiveWorkingHours': totalEffectiveWorkingHours,
      'totalOvertimeHours': totalOvertimeHours,
      'rawPresentDays': rawPresentDays,
      'presentDays': presentDays,
      'absentDays': absentDays,
      'prorationFactor': prorationFactor,
      'basicPay': basicPay,
      'hraAmount': hraAmount,
      'medicalAllowance': medicalAllowance,
      'conveyanceAllowance': conveyanceAllowance,
      'uniformAndSafetyAllowance': uniformAndSafetyAllowance,
      'bonus': bonus,
      'foodAllowance': foodAllowance,
      'specialAllowance': specialAllowance,
      'overtimePay': overtimePay,
      'totalSpecialAllowance': totalSpecialAllowance,
      'otherIncentive': otherIncentive,
      'epfEmployerEarnings': epfEmployerEarnings,
      'totalEarnings': totalEarnings,
      'epfEmployeeDeduction': epfEmployeeDeduction,
      'epfEmployerDeduction': epfEmployerDeduction,
      'esicDeduction': esicDeduction,
      'professionalTax': professionalTax,
      'tds': tds,
      'advanceSalaryRecovery': advanceSalaryRecovery,
      'loanRecovery': loanRecovery,
      'fuelAdvanceRecovery': fuelAdvanceRecovery,
      'otherDeductions': otherDeductions,
      'totalStatutoryDeductions': totalStatutoryDeductions,
      'totalOtherDeductions': totalOtherDeductions,
      'totalDeductions': totalDeductions,
      'netSalary': netSalary,
      'dailyRate': dailyRate,
      'hourlyRate': hourlyRate,
      'overtimeRate': overtimeRate,
      'status': status,
      'generatedBy': generatedBy,
      'generatedAt': generatedAt?.toIso8601String(),
      'notes': notes,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }
}

class SalaryCalculationRequest {
  final String employeeId;
  final int year;
  final int month;
  final double? tds;
  final double? advanceSalaryRecovery;
  final double? loanRecovery;
  final double? fuelAdvanceRecovery;
  final double? otherDeductions;
  final double? otherIncentive;
  final String? notes;

  SalaryCalculationRequest({
    required this.employeeId,
    required this.year,
    required this.month,
    this.tds,
    this.advanceSalaryRecovery,
    this.loanRecovery,
    this.fuelAdvanceRecovery,
    this.otherDeductions,
    this.otherIncentive,
    this.notes,
  });

  Map<String, dynamic> toJson() {
    return {
      'employeeId': employeeId,
      'year': year,
      'month': month,
      'tds': tds ?? 0.0,
      'advanceSalaryRecovery': advanceSalaryRecovery ?? 0.0,
      'loanRecovery': loanRecovery ?? 0.0,
      'fuelAdvanceRecovery': fuelAdvanceRecovery ?? 0.0,
      'otherDeductions': otherDeductions ?? 0.0,
      'otherIncentive': otherIncentive,
      'notes': notes,
    };
  }
}

File: civildesk-backend/civildesk-backend/src/main/java/com/civiltech/civildesk_backend/model/SalarySlip.java
package com.civiltech.civildesk_backend.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import java.time.YearMonth;

@Entity
@Table(name = "salary_slips")
@Data
@EqualsAndHashCode(callSuper = true)
@NoArgsConstructor
@AllArgsConstructor
public class SalarySlip extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "employee_id", nullable = false)
    private Employee employee;

    @Column(name = "year", nullable = false)
    private Integer year;

    @Column(name = "month", nullable = false)
    private Integer month; // 1-12

    // Calendar & Working Days
    @Column(name = "total_days_in_month")
    private Integer totalDaysInMonth;

    @Column(name = "working_days")
    private Integer workingDays; // Monday to Saturday

    @Column(name = "weekly_offs")
    private Integer weeklyOffs; // Sundays

    // Attendance Data
    @Column(name = "total_effective_working_hours")
    private Double totalEffectiveWorkingHours;

    @Column(name = "total_overtime_hours")
    private Double totalOvertimeHours;

    @Column(name = "raw_present_days")
    private Double rawPresentDays;

    @Column(name = "present_days")
    private Integer presentDays;

    @Column(name = "absent_days")
    private Integer absentDays;

    @Column(name = "proration_factor")
    private Double prorationFactor;

    // Earnings (Prorated)
    @Column(name = "basic_pay")
    private Double basicPay;

    @Column(name = "hra_amount")
    private Double hraAmount;

    @Column(name = "medical_allowance")
    private Double medicalAllowance;

    @Column(name = "conveyance_allowance")
    private Double conveyanceAllowance;

    @Column(name = "uniform_and_safety_allowance")
    private Double uniformAndSafetyAllowance;

    @Column(name = "bonus")
    private Double bonus;

    @Column(name = "food_allowance")
    private Double foodAllowance;

    @Column(name = "special_allowance")
    private Double specialAllowance; // Prorated special allowance

    @Column(name = "overtime_pay")
    private Double overtimePay;

    @Column(name = "total_special_allowance")
    private Double totalSpecialAllowance; // Special Allowance + Overtime Pay

    @Column(name = "other_incentive")
    private Double otherIncentive; // NOT prorated

    @Column(name = "epf_employer_earnings")
    private Double epfEmployerEarnings;

    @Column(name = "total_earnings")
    private Double totalEarnings;

    // Deductions
    @Column(name = "epf_employee_deduction")
    private Double epfEmployeeDeduction;

    @Column(name = "epf_employer_deduction")
    private Double epfEmployerDeduction;

    @Column(name = "esic_deduction")
    private Double esicDeduction;

    @Column(name = "professional_tax")
    private Double professionalTax;

    @Column(name = "tds")
    private Double tds;

    @Column(name = "advance_salary_recovery")
    private Double advanceSalaryRecovery;

    @Column(name = "loan_recovery")
    private Double loanRecovery;

    @Column(name = "fuel_advance_recovery")
    private Double fuelAdvanceRecovery;

    @Column(name = "other_deductions")
    private Double otherDeductions;

    @Column(name = "total_statutory_deductions")
    private Double totalStatutoryDeductions;

    @Column(name = "total_other_deductions")
    private Double totalOtherDeductions;

    @Column(name = "total_deductions")
    private Double totalDeductions;

    // Net Salary
    @Column(name = "net_salary")
    private Double netSalary;

    // Rates
    @Column(name = "daily_rate")
    private Double dailyRate;

    @Column(name = "hourly_rate")
    private Double hourlyRate;

    @Column(name = "overtime_rate")
    private Double overtimeRate;

    // Status
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private SalarySlipStatus status = SalarySlipStatus.DRAFT;

    @Column(name = "generated_by")
    private Long generatedBy; // User ID who generated the slip

    @Column(name = "generated_at")
    private java.time.LocalDateTime generatedAt;

    @Column(name = "notes", columnDefinition = "TEXT")
    private String notes;

    public enum SalarySlipStatus {
        DRAFT, FINALIZED, PAID, CANCELLED
    }

    // Helper method to get YearMonth
    public YearMonth getYearMonth() {
        return YearMonth.of(year, month);
    }

    // Helper method to get period string
    public String getPeriodString() {
        return String.format("%s %d", getMonthName(), year);
    }

    private String getMonthName() {
        String[] monthNames = {"", "January", "February", "March", "April", "May", "June",
                "July", "August", "September", "October", "November", "December"};
        return monthNames[month];
    }
}

